Sync/SyncTelemetry.swift

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Account
+
+public enum SyncReason: String {
+    case startup = "startup"
+    case scheduled = "scheduled"
+    case backgrounded = "backgrounded"
+    case user = "user"
+    case syncNow = "syncNow"
+    case didLogin = "didLogin"
+}
+
+public protocol Stats {
+    func hasData() -> Bool
+}
+
+public struct SyncUploadStats: Stats {
+    var sent: Int = 0
+    var sentFailed: Int = 0
+
+    public func hasData() -> Bool {
+        return sent > 0 || sentFailed > 0
+    }
+}
+
+public struct SyncDownloadStats: Stats {
+    var applied: Int = 0
+    var succeeded: Int = 0
+    var failed: Int = 0
+    var newFailed: Int = 0
+    var reconciled: Int = 0
+
+    public func hasData() -> Bool {
+        return applied > 0 ||
+               succeeded > 0 ||
+               failed > 0 ||
+               newFailed > 0 ||
+               reconciled > 0
+    }
+}
+
+// TODO(sleroux): Implement various bookmark validation issues we can run into
+public struct ValidationStats: Stats {
+    public func hasData() -> Bool {
+        return false
+    }
+}
+
+public class StatsSession {
+    private var took: UInt64 = 0
+    private var startTime: Timestamp?
+
+    public func start(time: Timestamp = Date.now()) {
+        self.startTime = time
+    }
+
+    public func hasStarted() -> Bool {
+        return startTime != nil
+    }
+
+    public func end() -> Self {
+        guard let startTime = startTime else {
+            assertionFailure("SyncOperationStats called end without first calling start!")
+            return self
+        }
+
+        took = Date.now() - startTime
```

Comment:
Note that at this point if the wall clock has jumped backwards, Firefox will crash. Be very, very careful when using `UInt64`!